File: 367_Valid_Perfect_Square.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/valid-perfect-square/
# Given a positive integer num, write a function which returns True if num is a perfect square else False.

# Binary search between 1 and num. Test if middle of earch region is the square root of num. If not, search region
# to left or right until solution or no more search region.
# Time - O(log n)
# Space - O(1)

class Solution(object):
    def isPerfectSquare(self, num):
        """
        :type num: int
        :rtype: bool
        """
        left, right = 1, num

        while left <= right:

            mid = (left + right) // 2
            square = mid * mid

            if square == num:
                return True

            if square > num:
                right = mid - 1
            else:
                left = mid + 1

        return False

File: java/367_Valid_Perfect_Square.java
/*
https://leetcode.com/problems/logger-rate-limiter/
Given a positive integer num, write a function which returns True if num is a perfect square else False.

Binary search the range from 1 to n inclusive.  Use long for mid to avoid overflow with mid * mid.
Return false if left and right are inverted.
Time - O(log n)
Space - O(1)
*/

public class Solution {
    public boolean isPerfectSquare(int num) {
        long left = 1;
        long right = num;

        while (left <= right) {
            long mid = left + (right - left) / 2;

            if (mid * mid == num)
                return true;
            else if (mid * mid > num)   // mid too big, sqrt(num) must be LHS of mid
                right = mid - 1;
            else
                left = mid + 1;
        }

        return false;
    }
}
